spec/consensus/light-client.md

```diff
@@ -60,259 +61,368 @@ For the purpose of this lite client specification, we assume that the Tendermint
+**VerifyHeader.** The function *VerifyHeader* captures high level logic, i.e., application call to the lite client module to (optionally download) and
+verify header for some height. The core verification logic is captured by *CanTrust* function that iteratively try to establish trust in given header
+by relying on *CheckSupport* function.
 
-Towards Lite Client Completeness:
 
-- The check is successful if sufficiently many validators of _h1_ are still validators in _h2_ and signed _h2_.
-- If _h2.Header.height = h1.Header.height + 1_, and both headers were generated correctly, the test passes
+```go
+func VerifyHeader(height, trustThreshold) error {
+  if h2, exists := Store.Get(height); exists {
+    if isWithinTrustedPeriod(h2) { return nil }
+    return ErrHeaderNotWithinTrustedPeriod(h2)
+  }
 
-_Verification Condition:_ We may need a Tendermint invariant stating that if _h2.Header.height = h1.Header.height + 1_ then _signers(h2.Commit) \subseteq h1.Header.NextV_.
+  h2 := Commit(height)
+  if !verify(h2) { return ErrInvalidHeader(h2) }
+  if !isWithinTrustedPeriod(h2) { return ErrHeaderNotWithinTrustedPeriod(h2) }
 
-_Remark_: The variable _trustlevel_ can be used if the user believes that relying on one correct validator is not sufficient. However, in case of (frequent) changes in the validator set, the higher the _trustlevel_ is chosen, the more unlikely it becomes that CheckSupport returns true for non-adjacent headers.
+  // get the highest trusted headers lower than h2
+  h1 = Store.HighestTrustedSmallerThan(height)
+  if h1 == nil { return ErrNoTrustedHeader }
 
-**Bisection.** The following function uses CheckSupport in a recursion to find intermediate headers that allow to establish a sequence of trust.
+  err = CanTrust(h1, h2, trustThreshold)  // or CanTrustBisection((h1, h2, trustThreshold)
+  if err != nil { return err }
 
-```go
-func Bisection(h1,h2,trustlevel) bool{
-  if CheckSupport(h1,h2,trustlevel) {
-    return true
+  if isWithinTrustedPeriod(h2) {
+    Store.add(h2)
+    // we store only trusted headers, as we assume that only trusted headers
+    // are influencing end user business decisions.
+    return nil
   }
-  if h2.Header.height == h1.Header.height + 1 {
-    // we have adjacent headers that are not matching (failed
-    // the CheckSupport)
-    // we could submit evidence here
-    return false
+  return ErrHeaderNotTrusted(h2)
+}
+
+
+// return nil in case we can trust header h2 based on header h1; otherwise return error
+// where error captures the nature of the error.
+func CanTrust(h1,h2,trustThreshold) error {
+  assume h1.Header.Height < h2.header.Height
+
+  th := h1 // th is trusted header
+  untrustedHeaders := [h2]
+
+  while true {
+    for h in untrustedHeaders {
+        // we assume here that iteration is done in the order of header heights
```

Comment:
it is more efficient if we loop in descending height order. And shows the greedy approach of checking support with high header first.
Also reads better if we do the "return" checks first
```
    for h in reverse(untrustedHeaders) {
        // we assume here that iteration is done in the reverse order of header heights
        err = CheckSupport(th,h,trustThreshold)
        if (err != ErrTooMuchChange) { return err }
        if err == nil {
            if h == h2 { return nil }
            th = h
            Store.Add(th)
            untrustedHeaders.RemoveHeadersSmallerOrEqual(th.Header.Height)
        }
    }
```
This way once `RemoveHeadersSmallerOrEqual` is called once, `untrustedHeaders` is empty and the `for` loop finishes. 
Also it's nice if we always call `Store.Add(th)` to show that we only store trusted headers.

Reply:
With Zarko we discussed that once we get a new (higher) trusted header, we can directly check whether the new header is sufficient to validate the target header. This is different from what is encoded here, and corresponds to the reverse order you propose.

Back then we had the impression that what is more efficient depends a lot on the changes in the validator sets. If the validator set of the new trusted header is the same as the initial header, I guess the current version is better. If there is already a big change, the reverse order might be better. We might think about checking the change in the validator set and adapt the search strategy accordingly. However, I am not sure whether this would give better guarantees (theoretical or practical).

If I understand correctly, what is encoded now will do the same sequence of checks as the recursive bisection. Is this right? 

I guess untrustedHeaders is somewhat similar to the call stack implemented by Igor in the TLA+ stack of the recursive bisection.

Reply:
> With Zarko we discussed that once we get a new (higher) trusted header, we can directly check whether the new header is sufficient to validate the target header. This is different from what is encoded here, and corresponds to the reverse order you propose.

Yes, I propose we stop at the highest header that checks support by going from highest to lowest and stopping at first success. Zarko's code goes from lowest to highest. The differences are:
high->low: finish faster, check and save less headers. The latter could be actually a disatvantage depending on client use cases.
low->high: finish later, check and save more headers. So it could be a good thing to have more recent trusted headers as the lower one expire first.

> We might think about checking the change in the validator set and adapt the search strategy accordingly.

I have to find some old code I had that retrieves validators sets from cosmos-hub and computes the total voting power at each block. We can do some prototype implementations with different versions.

> If I understand correctly, what is encoded now will do the same sequence of checks as the recursive bisection. Is this right?

Yes. My proposal actually checks less headers so not same seq. Thinking about it more, I think it is better do what is now in the spec and discard my comment.

> I guess untrustedHeaders is somewhat similar to the call stack implemented by Igor in the TLA+ stack of the recursive bisection.

Yes, it is exactly that.